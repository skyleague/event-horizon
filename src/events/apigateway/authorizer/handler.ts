import type { Try } from '@skyleague/axioms'
import { mapTry, recoverTry, tryToError } from '@skyleague/axioms'
import type { APIGatewayIAMAuthorizerResult, APIGatewaySimpleAuthorizerResult } from 'aws-lambda'
import type { APIGatewayRequestAuthorizerEventV2Schema } from '../../../aws/http.js'
import type { APIGatewayRequestAuthorizerEventSchema } from '../../../aws/rest.js'
import type { MaybeGenericParser } from '../../../parsers/types.js'
import { ioLoggerChild } from '../../functions/io-logger-child.js'
import type { LambdaContext } from '../../types.js'
import type { GatewayVersion } from '../event/types.js'
import { authorizerErrorHandler } from './functions/error-handler.js'
import { authorizerIOLogger } from './functions/io-logger.js'
import { authorizerIOValidate } from './functions/io-validate.js'
import { authorizerParseEvent } from './functions/parse-event.js'
import { authorizerSerializeResponse } from './functions/serialize-response.js'
import type { RequestAuthorizerHandler, SecuritySchemes } from './types.js'

export async function handleAuthorizerEvent<
    Configuration,
    Service,
    Profile extends MaybeGenericParser,
    Path extends MaybeGenericParser,
    Query extends MaybeGenericParser,
    Headers extends MaybeGenericParser,
    Context extends MaybeGenericParser,
    Security extends SecuritySchemes | undefined,
    GV extends GatewayVersion,
>(
    handler: RequestAuthorizerHandler<Configuration, Service, Profile, Path, Query, Headers, Context, Security, GV>,
    event: APIGatewayRequestAuthorizerEventV2Schema | APIGatewayRequestAuthorizerEventSchema,
    context: LambdaContext<Configuration, Service, Profile>,
): Promise<Try<APIGatewayIAMAuthorizerResult | APIGatewaySimpleAuthorizerResult>> {
    const { request } = handler
    const parseEventFn = authorizerParseEvent(request)
    const ioValidateFn = authorizerIOValidate(request)

    const serializeResponseFn = authorizerSerializeResponse()
    const errorHandlerFn = authorizerErrorHandler(context)
    const inputOutputFn = authorizerIOLogger(context)
    const ioLoggerChildFn = ioLoggerChild(context, context.logger)

    const unvalidatedAuthorizerEvent = mapTry(event, (e) => parseEventFn.before(e))

    const authorizerEvent = await mapTry(unvalidatedAuthorizerEvent, (e) => {
        ioLoggerChildFn.before({
            ...e.path,
        })

        return ioValidateFn.before(e)
    })

    inputOutputFn.before(authorizerEvent)

    const transform = await mapTry(authorizerEvent, (e) => request.handler(e, context))
    const response = recoverTry(transform, (f) => errorHandlerFn.onError(f))

    inputOutputFn.after(response)
    ioLoggerChildFn.after()

    return tryToError(mapTry(response, (r) => serializeResponseFn.onAfter(event, r)))
}
