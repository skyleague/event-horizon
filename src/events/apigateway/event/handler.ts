import type { Try } from '@skyleague/axioms'
import { mapTry, recoverTry } from '@skyleague/axioms'
import type { APIGatewayProxyResult, APIGatewayProxyResultV2 } from 'aws-lambda'
import type { APIGatewayProxyEventV2Schema } from '../../../aws/apigateway/http.type.js'
import type { APIGatewayProxyEventSchema } from '../../../aws/apigateway/rest.type.js'
import type { MaybeGenericParser } from '../../../parsers/types.js'
import { ioLoggerChild } from '../../functions/io-logger-child.js'
import type { LambdaContext, ResponseFromHandler } from '../../types.js'
import type { SecurityRequirements } from '../types.js'
import { httpErrorHandler } from './functions/error-handler.js'
import { httpIOLogger } from './functions/io-logger.js'
import { httpIOValidate } from './functions/io-validate.js'
import { httpParseEvent } from './functions/parse-event.js'
import { httpSerializeResponse } from './functions/serialize-response.js'
import type { AuthorizerSchema, HTTPEventHandler, HTTPHandler, HTTPResponse, Responses } from './types.js'
import type { GatewayVersion } from './types.js'

export async function handleHTTPEvent<
    Configuration,
    Service,
    Profile extends MaybeGenericParser,
    Body extends MaybeGenericParser,
    Path extends MaybeGenericParser,
    Query extends MaybeGenericParser,
    Headers extends MaybeGenericParser,
    Result extends Responses,
    Security extends SecurityRequirements | undefined,
    GV extends GatewayVersion,
    Authorizer extends AuthorizerSchema<GV>,
>(
    handler: HTTPHandler<Configuration, Service, Profile, Body, Path, Query, Headers, Result, Security, GV, Authorizer>,
    event: APIGatewayProxyEventV2Schema | APIGatewayProxyEventSchema,
    context: LambdaContext<Configuration, Service, Profile>,
): Promise<Try<APIGatewayProxyResult | APIGatewayProxyResultV2>> {
    const { http } = handler
    const parseEventFn = httpParseEvent(http)
    const ioValidateFn = httpIOValidate(http)
    const serializeResponseFn = httpSerializeResponse()
    const errorHandlerFn = httpErrorHandler(context, handler.serialize?.error)
    const inputOutputFn = httpIOLogger(http, context)
    const ioLoggerChildFn = ioLoggerChild(context, context.logger)

    const unvalidatedAuthorizer = mapTry(event, (e) => parseEventFn.authorization(e))
    const unvalidatedHttpEvent = mapTry(unvalidatedAuthorizer, (a) => parseEventFn.before(event, a))

    const httpEvent = await mapTry(unvalidatedHttpEvent, (e) => {
        ioLoggerChildFn.before({
            ...e.path,
        })

        return ioValidateFn.before(e)
    })

    inputOutputFn.before(httpEvent)

    const transform = await mapTry(httpEvent, (e) => http.handler(e, context))
    const response = recoverTry(transform, (f) => errorHandlerFn.onError(f))

    inputOutputFn.after(
        // @todo remove cast
        response as Awaited<
            ResponseFromHandler<
                HTTPEventHandler<Configuration, Service, Profile, Body, Path, Query, Headers, Result, Security, GV, Authorizer>
            >
        >,
    )
    ioLoggerChildFn.after()

    // @todo remove cast
    return mapTry(response, (r) => serializeResponseFn.onAfter(r as HTTPResponse))
}
