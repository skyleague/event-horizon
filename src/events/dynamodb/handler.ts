import type { Try } from '@skyleague/axioms'
import { isLeft, mapLeft, mapTry, tryToEither } from '@skyleague/axioms'
import type { DynamoDBBatchItemFailure, DynamoDBBatchResponse } from 'aws-lambda/trigger/dynamodb-stream.js'
import type { DynamoDBStreamRecord } from '../../dev/aws/dynamodb/dynamodb.type.js'
import { ioLoggerChild } from '../functions/io-logger-child.js'
import { ioLogger } from '../functions/io-logger.js'
import type { LambdaContext } from '../types.js'
import { dynamodbErrorHandler } from './functions/error-handler.js'
import { dynamodbParseEvent } from './functions/parse-event.js'
import type { DynamoDBStreamHandler } from './types.js'

export async function handleDynamoDBStreamEvent<Configuration, Service, Profile>(
    handler: DynamoDBStreamHandler<Configuration, Service, Profile>,
    events: DynamoDBStreamRecord[],
    context: LambdaContext<Configuration, Service, Profile>,
    // biome-ignore lint/suspicious/noConfusingVoidType: this is the real type we want here
): Promise<Try<DynamoDBBatchResponse | void>> {
    const { dynamodb } = handler
    const errorHandlerFn = dynamodbErrorHandler(context)
    const parseEventFn = dynamodbParseEvent()
    const ioLoggerFn = ioLogger({ type: 'dynamodb' }, context)
    const ioLoggerChildFn = ioLoggerChild(context, context.logger)

    let failures: DynamoDBBatchItemFailure[] | undefined = undefined

    for (const [i, event] of events.entries()) {
        const item = { item: i }

        const dynamodbEvent = mapTry(event, (e) => {
            const unvalidatedDynamodbEvent = parseEventFn.before(e)

            ioLoggerChildFn.before({
                eventId: unvalidatedDynamodbEvent.raw.eventID,
            })

            return unvalidatedDynamodbEvent
        })

        ioLoggerFn.before(dynamodbEvent, item)

        const transformed = await mapTry(dynamodbEvent, (success) => dynamodb.handler(success, context))

        const eitherTransformed = tryToEither(transformed)
        const response = mapLeft(eitherTransformed, (e) => errorHandlerFn.onError(event, e))

        ioLoggerFn.after(undefined, item)
        ioLoggerChildFn.after()

        if (isLeft(response)) {
            failures ??= []
            failures.push(response.left)
        }
    }
    if (failures !== undefined) {
        return { batchItemFailures: failures }
    }
}
