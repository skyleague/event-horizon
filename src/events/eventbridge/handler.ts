import { eventBridgeParseEvent } from './functions/parse-event.js'
import type { EventBridgeEvent, EventBridgeHandler } from './types.js'

import { ioLogger } from '../functions/io-logger.js'
import { ioValidate } from '../functions/io-validate.js'
import type { DefaultServices, LambdaContext } from '../types.js'

import type { Try } from '@skyleague/axioms'
import { mapTry } from '@skyleague/axioms'
import type { UndefinedOnPartialDeep } from '@skyleague/axioms/types'
import type { EventBridgeEvent as AWSEventBridgeEvent } from 'aws-lambda'

export async function handleEventBridgeEvent<
    Configuration,
    Service extends DefaultServices | undefined,
    Profile,
    Payload,
    Result,
>(
    handler: EventBridgeHandler<Configuration, Service, Profile, Payload, Result>,
    event: UndefinedOnPartialDeep<AWSEventBridgeEvent<string, unknown>>,
    context: LambdaContext<Configuration, Service, Profile>,
): Promise<Try<unknown>> {
    const { eventBridge } = handler

    const parseEventFn = eventBridgeParseEvent()
    const ioValidateFn = ioValidate<EventBridgeEvent>()
    const ioLoggerFn = ioLogger({ type: 'eventbridge' }, context)

    const ebEvent = mapTry(event, (e): EventBridgeEvent<Payload> => {
        const unvalidatedEbEvent = parseEventFn.before(e)
        return ioValidateFn.before<Payload, 'payload', EventBridgeEvent<Payload>>(
            eventBridge.schema.payload,
            unvalidatedEbEvent,
            'payload',
        )
    })

    ioLoggerFn.before(ebEvent)

    const unvalidatedResponse = await mapTry(ebEvent, (e) => eventBridge.handler(e, context))
    const response = mapTry(unvalidatedResponse, (t) => ioValidateFn.after(eventBridge.schema.result, t))

    ioLoggerFn.after(response)

    return response
}
