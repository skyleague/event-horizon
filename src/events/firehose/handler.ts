import { firehoseErrorHandler } from './functions/error-handler.js'
import { firehoseParseEvent } from './functions/parse-event.js'
import { firehoseSerializeTransformation } from './functions/serialize-transformation.js'
import type { FirehoseTransformationEvent, FirehoseTransformationResult } from './types.js'
import type { FirehoseTransformationHandler } from './types.js'

import { ioLogger } from '../functions/io-logger.js'
import { ioValidate } from '../functions/io-validate.js'
import type { DefaultServices, LambdaContext } from '../types.js'

import type { Try } from '@skyleague/axioms'
import { isSuccess, mapTry, transformTry } from '@skyleague/axioms'
import type {
    FirehoseTransformationResult as AWSFirehoseTransformationResult,
    FirehoseTransformationResultRecord,
} from 'aws-lambda'
import type { KinesisFirehoseRecord } from '../../dev/aws/firehose/firehose.type.js'

export async function handleFirehoseTransformation<
    const Configuration,
    const Service extends DefaultServices | undefined,
    const Profile,
    const Payload,
    const Result,
>(
    handler: FirehoseTransformationHandler<Configuration, Service, Profile, Payload, Result>,
    events: KinesisFirehoseRecord[],
    context: LambdaContext<Configuration, Service, Profile>,
): Promise<Try<AWSFirehoseTransformationResult>> {
    const { firehose } = handler

    const serializeTransformationFn = firehoseSerializeTransformation()
    const errorHandlerFn = firehoseErrorHandler(context)
    const parseEventFn = firehoseParseEvent(firehose)
    const ioValidateFn = ioValidate<FirehoseTransformationEvent, FirehoseTransformationResult>()
    const ioLoggerFn = ioLogger({ type: 'firehose' }, context)

    const responses: FirehoseTransformationResultRecord[] = []
    for (const [i, event] of events.entries()) {
        const item = { item: i }

        const firehoseEvent = mapTry(event, (e) => {
            const unvalidatedFirehoseEvent = parseEventFn.before(e)
            return ioValidateFn.before(firehose.schema.payload, unvalidatedFirehoseEvent, 'payload')
        })

        ioLoggerFn.before(firehoseEvent, item)

        const unvalidatedTransformed = await mapTry(firehoseEvent, (success) => firehose.handler(success, context))
        const transformed = mapTry(unvalidatedTransformed, (t) => ioValidateFn.after(firehose.schema.result, t, 'payload'))

        const response = transformTry(
            transformed,
            (x) => serializeTransformationFn.onAfter(event, x),
            (e) => errorHandlerFn.onError(event, e),
        )

        ioLoggerFn.after(undefined, item)

        if (isSuccess(response)) {
            responses.push(response)
        } else {
            return response
        }
    }

    return { records: responses }
}
