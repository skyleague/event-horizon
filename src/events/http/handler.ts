import { httpErrorHandler } from './functions/error-handler.js'
import { httpIOLogger } from './functions/io-logger.js'
import { httpIOValidate } from './functions/io-validate.js'
import { httpParseEvent } from './functions/parse-event.js'
import { httpSerializeResponse } from './functions/serialize-response.js'
import type { GatewayVersion, HTTPHandler } from './types.js'

import { ioLoggerChild } from '../functions/io-logger-child.js'
import type { LambdaContext } from '../types.js'

import type { Try } from '@skyleague/axioms'
import { mapTry, recoverTry } from '@skyleague/axioms'
import type { APIGatewayProxyResult, APIGatewayProxyResultV2 } from 'aws-lambda'
import type { APIGatewayProxyEventV2Schema } from '../../aws/apigateway/http.type.js'
import type { APIGatewayProxyEventSchema } from '../../aws/apigateway/rest.type.js'

export async function handleHTTPEvent<
    Configuration,
    Service,
    Profile,
    Body,
    Path,
    Query,
    Headers,
    Result,
    GV extends GatewayVersion = 'http',
>(
    handler: HTTPHandler<Configuration, Service, Profile, Body, Path, Query, Headers, Result, GV>,
    event: APIGatewayProxyEventV2Schema | APIGatewayProxyEventSchema,
    context: LambdaContext<Configuration, Service, Profile>,
): Promise<Try<APIGatewayProxyResult | APIGatewayProxyResultV2>> {
    const { http } = handler
    const parseEventFn = httpParseEvent(http)
    const ioValidateFn = httpIOValidate<Configuration, Service, Profile, Body, Path, Query, Headers, Result, GV>()
    const serializeResponseFn = httpSerializeResponse()
    const errorHandlerFn = httpErrorHandler(context, handler.serialize?.error)
    const inputOutputFn = httpIOLogger(http, context)
    const ioLoggerChildFn = ioLoggerChild(context, context.logger)

    const unvalidatedHttpEvent = mapTry(event, (e) => parseEventFn.before(e))

    const httpEvent = mapTry(unvalidatedHttpEvent, (e) => {
        ioLoggerChildFn.before({
            ...e.path,
        })

        return ioValidateFn.before(http, e)
    })

    inputOutputFn.before(httpEvent)

    const transform = await mapTry(httpEvent, (e) => http.handler(e, context))
    const response = recoverTry(transform, (f) => errorHandlerFn.onError(f))

    inputOutputFn.after(response)
    ioLoggerChildFn.after()

    return mapTry(response, (r) => serializeResponseFn.onAfter(r))
}
