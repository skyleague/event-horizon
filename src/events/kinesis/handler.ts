import type { Try } from '@skyleague/axioms'
import { isLeft, mapLeft, mapTry, tryAsValue, tryToEither } from '@skyleague/axioms'
import type { KinesisStreamBatchItemFailure, KinesisStreamBatchResponse } from 'aws-lambda'
import type { KinesisDataStreamRecord } from '../../aws/kinesis/kinesis.type.js'
import { ioLoggerChild } from '../functions/io-logger-child.js'
import { ioLogger } from '../functions/io-logger.js'
import { ioValidate } from '../functions/io-validate.js'
import type { LambdaContext } from '../types.js'
import { kinesisErrorHandler } from './functions/error-handler.js'
import { kinesisParseEvent } from './functions/parse-event.js'
import type { KinesisEvent, KinesisHandler } from './types.js'

export async function handleKinesisEvent<Configuration, Service, Profile, Payload>(
    handler: KinesisHandler<Configuration, Service, Profile, Payload>,
    events: KinesisDataStreamRecord[],
    context: LambdaContext<Configuration, Service, Profile>,
    // biome-ignore lint/suspicious/noConfusingVoidType: this is the real type we want here
): Promise<Try<KinesisStreamBatchResponse | void>> {
    const { kinesis } = handler

    const errorHandlerFn = kinesisErrorHandler(context)
    const parseEventFn = kinesisParseEvent(kinesis)
    const ioValidateFn = ioValidate<KinesisEvent>()
    const ioLoggerFn = ioLogger({ type: 'kinesis' }, context)
    const ioLoggerChildFn = ioLoggerChild(context, context.logger)

    let failures: KinesisStreamBatchItemFailure[] | undefined = undefined

    for (const [i, event] of events.entries()) {
        const item = { item: i }

        const kinesisEvent = mapTry(event, (e) => {
            const unvalidatedKinesisEvent = parseEventFn.before(e)

            ioLoggerChildFn.before({
                eventId: unvalidatedKinesisEvent.raw.eventID,
            })

            return ioValidateFn.before(kinesis.schema.payload, unvalidatedKinesisEvent, 'payload')
        })

        ioLoggerFn.before(tryAsValue(kinesisEvent), item)

        const transformed = await mapTry(kinesisEvent, (e) => kinesis.handler(e, context))

        const response = mapLeft(tryToEither(transformed), (e) => errorHandlerFn.onError(event, e))

        if (isLeft(response)) {
            ioLoggerFn.after(response.left, item)
            failures ??= []
            failures.push(response.left)
        } else {
            ioLoggerFn.after(undefined, item)
        }

        ioLoggerChildFn.after()
    }
    if (failures !== undefined) {
        return { batchItemFailures: failures }
    }
}
