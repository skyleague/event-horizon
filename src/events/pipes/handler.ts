import { pipeErrorHandler } from './functions/error-handler.js'
import { pipeParseEvent } from './functions/parse-event.js'
import type { PipesHandler } from './types.js'

import { ioLogger } from '../functions/io-logger.js'
import { ioValidate } from '../functions/io-validate.js'
import type { LambdaContext } from '../types.js'

import type { Try } from '@skyleague/axioms'
import { eitherAsValue, enumerate, isLeft, mapLeft, mapTry, tryToEither } from '@skyleague/axioms'

export async function handlePipesEvent(
    handler: PipesHandler,
    events: unknown[],
    context: LambdaContext
): Promise<Try<unknown[]>> {
    const { pipe } = handler
    const errorHandlerFn = pipeErrorHandler(context)
    const parseEventFn = pipeParseEvent(pipe)
    const ioValidateFn = ioValidate()
    const ioLoggerFn = ioLogger({ type: 'pipe' }, context)

    ioLoggerFn.before(events)
    const responses: unknown[] = []
    for (const [i, event] of enumerate(events)) {
        const item = { item: i }

        const pipesEvent = mapTry(event, (e) => {
            const unvalidatedPipesEvent = parseEventFn.before(e)
            return ioValidateFn.before(pipe.schema.payload, unvalidatedPipesEvent)
        })

        ioLoggerFn.before(pipesEvent, item)

        const transformed = await mapTry(pipesEvent, (success) => pipe.handler(success, context))

        const eitherTransformed = tryToEither(transformed)
        const response = mapLeft(eitherTransformed, (e) => errorHandlerFn.onError(event, e))

        ioLoggerFn.after(eitherAsValue(response), item)

        if (isLeft(response)) {
            return response.left
        } else {
            responses.push(response.right)
        }
    }

    return responses
}
