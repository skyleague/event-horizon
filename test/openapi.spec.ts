import { expect, it } from 'vitest'
import { handlerJson } from '../examples/pet-store/functions/openapi/handler.js'

it('buids the petstore openapi', () => {
    expect(handlerJson.http.handler()).toMatchSnapshot()
})
