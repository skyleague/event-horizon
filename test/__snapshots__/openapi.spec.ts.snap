// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buids the petstore openapi 1`] = `
{
  "body": {
    "components": {
      "parameters": {
        "Status": {
          "$ref": "#/components/schemas/Status",
        },
      },
      "requestBodies": {
        "Pet": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/Pet",
              },
            },
          },
          "description": "Pet object from the store",
        },
      },
      "responses": {
        "Error": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/Error",
              },
            },
          },
          "description": "The default error error response for both 400 & 500 type errors",
        },
        "Pet": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/Pet",
              },
            },
          },
          "description": "Pet object from the store",
        },
        "PetArray": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/PetArray",
              },
            },
          },
          "description": "A list of Pet objects",
        },
      },
      "schemas": {
        "Category": {
          "additionalProperties": true,
          "properties": {
            "id": {
              "type": "integer",
            },
            "name": {
              "type": "string",
            },
          },
          "required": [
            "id",
            "name",
          ],
          "title": "Category",
          "type": "object",
        },
        "Error": {
          "description": "The default error error response for both 400 & 500 type errors",
          "properties": {
            "message": {
              "description": "A detailed message of the error.",
              "type": "string",
            },
            "stack": {
              "description": "When debugging is enabled, a stack trace is exposed.",
              "type": "string",
            },
            "statusCode": {
              "description": "The status code of the response.",
              "type": "integer",
            },
          },
          "required": [
            "statusCode",
            "message",
          ],
          "title": "Default Error Response",
          "type": "object",
        },
        "Pet": {
          "additionalProperties": true,
          "description": "Pet object from the store",
          "properties": {
            "category": {
              "$ref": "#/components/schemas/Category",
            },
            "id": {
              "type": "integer",
            },
            "name": {
              "type": "string",
            },
            "photoUrls": {
              "items": {
                "type": "string",
              },
              "type": "array",
            },
            "status": {
              "enum": [
                "available",
                "pending",
                "sold",
              ],
            },
            "tags": {
              "items": {
                "$ref": "#/components/schemas/Tag",
              },
              "type": "array",
            },
          },
          "required": [
            "name",
            "photoUrls",
          ],
          "title": "Pet",
          "type": "object",
        },
        "PetArray": {
          "description": "A list of Pet objects",
          "items": {
            "$ref": "#/components/schemas/Pet",
          },
          "title": "PetArray",
          "type": "array",
        },
        "Status": {
          "enum": [
            "available",
            "pending",
            "sold",
          ],
          "title": "Status",
        },
        "Tag": {
          "additionalProperties": true,
          "properties": {
            "id": {
              "type": "integer",
            },
            "name": {
              "type": "string",
            },
          },
          "title": "Tag",
          "type": "object",
        },
      },
    },
    "info": {
      "title": "@skyleague/event-horizon",
      "version": "6.1.0",
    },
    "openapi": "3.0.1",
    "paths": {
      "/pet": {
        "post": {
          "description": "Add a new pet to the store",
          "parameters": [],
          "requestBody": {
            "$ref": "#/components/requestBodies/Pet",
          },
          "responses": {
            "200": {
              "$ref": "#/components/responses/Pet",
            },
            "default": {
              "$ref": "#/components/responses/Error",
            },
          },
        },
      },
      "/pet/findByStatus": {
        "get": {
          "description": "Multiple status values can be provided with comma separated strings",
          "parameters": [
            {
              "in": "query",
              "name": "status",
              "required": true,
              "schema": {
                "$ref": "#/components/parameters/Status",
              },
            },
          ],
          "responses": {
            "200": {
              "$ref": "#/components/responses/PetArray",
            },
            "default": {
              "$ref": "#/components/responses/Error",
            },
          },
          "summary": "Finds Pets by status",
        },
      },
    },
  },
  "statusCode": 200,
}
`;
